### DISCUSSION TITLE
分享自用的V5配置，可run，不知道逻辑上有没有错，请指教

### DISCUSSION THREAD

**Discussion by Mrican** [IrineSistiana/mosdns#625] [Ideas]
参考https://github.com/pmkol/easymosdns 和 https://github.com/IrineSistiana/mosdns-cn

感谢两位的付出！

# Powered by Mr.ican

log:
    level: error
    file: "./mosdns.log"

plugins:
    - tag: hosts
      type: hosts
      args:
        files: ./hosts.txt    #   https://cdn.jsdelivr.net/gh/VeleSila/yhosts/hosts.txt

    - tag: cache        # 缓存
      type: cache
      args:
        size: 1024
        lazy_cache_ttl: 86400

    - tag: adlist
      type: domain_set
      args:
        files: ./rules/ad_domain_list.txt

    - tag: china_domain
      type: domain_set
      args:
        files: ./rules/china_domain_list.txt

    - tag: cdncn_domain
      type: domain_set
      args:
        files: ./rules/cdn_domain_list.txt

    - tag: ecs_cn_domain
      type: domain_set
      args:
        files: ./rules/ecs_cn_domain.txt

    - tag: gfwlist
      type: domain_set
      args:
        files: ./rules/gfw_domain_list.txt

    - tag: local_ip        # china ip
      type: ip_set
      args:
        files: ./rules/china_ip_list.txt

     - tag: gfw_ip      # gfw ip
      type: ip_set
      args:
        files: ./rules/gfw_ip_list.txt

    - tag: forward_local
      type: forward
      args:
        upstreams:
            - addr: https://sm2.doh.pub/dns-query
              bootstrap: 119.29.29.29

    - tag: forward_remote
      type: forward
      args:
        #concurrent: 2
        upstreams:
            - addr: https://doh.apad.pro/dns-query
              bootstrap: 119.29.29.29
            #- addr: https://8.8.8.8/dns-query  太慢了，家里似乎访问不了

    - tag: local_sequence       # fallback的primary服务器,如果本地上游的应答包含 --local-ip 本地 IP。则直接采用本地上游的结果。结束。
      type: sequence
      args:
        - exec: $forward_local
        #- exec: ttl 60-3600
        - matches: "resp_ip $local_ip"
          exec: accept
        - matches: cname akamaiedge.net akadns.net edgekey.net
          exec: accept

    - tag: remote_sequence      # fallback的secondary服务器，否则采用远程上游的结果。结束。
      type: sequence
      args:
        - exec: prefer_ipv4
        - exec: $forward_remote
        #- exec: ttl 300-3600
        - matches: "!resp_ip $gfw_ip"
          exec: accept

    - tag: fallback      # fallback sequence
      type: fallback
      args:
        primary: local_sequence
        secondary: remote_sequence
        threshold: 500
        always_standby: false

    - tag: main_sequence
      type: sequence
      args:
        - exec: $hosts
        - matches: has_resp
          exec: accept         # 结束。

        - matches: qtype 12
          exec: reject 0            # reject 3 是屏蔽，reject 0是什么？没查到

        - matches: qtype 65
          exec: black_hole 127.0.0.1 ::1
        - matches: has_resp
          exec: accept         # 结束。   这里往上三行是不是可以用reject 3 代替？？

        - matches: qname $adlist
          exec: reject 3

        - exec: $cache      # 然后。查找 cache。
        - matches: has_resp  # 如果有应答了(上一步 cache 找到应答)
          exec: accept         # 结束。

        - matches: qname github.com
          exec: ecs 168.95.1.0/22 2001:b000:168::1/48      # 结束。   这里ipv6后面的/48是对吗？？
        - exec: $forward_remote
        #- exec: ttl 300-3600
        - matches: has_wanted_ans      #  has_resp 跟 has_wanted_ans 不太理解区别
          exec: accept         # 结束。

        - matches: qname $china_domain $cdncn_domain $ecs_cn_domain
          exec: $forward_local
        - matches: has_wanted_ans
          exec: accept         # 结束。

        - matches: qname $gfwlist
          exec: prefer_ipv4
        - exec: ecs 1.2.3.0/24 2001:dc7:1000::1/48
        - exec: $forward_remote
        - matches: has_wanted_ans
          exec: accept         # 结束。

        - exec: $fallback

    - tag: udp_server
      type: udp_server
      args:
        entry: main_sequence
        listen: 127.0.0.1:53

    - tag: tcp_server
      type: tcp_server
      args:
        entry: main_sequence
        listen: 127.0.0.1:53

**Comment by urlesistiana** [comment]
> reject 3 是屏蔽，reject 0是什么？没查到

reject 后面的是 dns rcode，不同 rcode 意思不一样，谷歌搜一下。

```yaml
    - matches: qtype 65
      exec: black_hole 127.0.0.1 ::1
```

qtype 65 的 请求是不能用 IP 屏蔽的。所以这段配置无意义。换 reject 0。

> has_resp 跟 has_wanted_ans 不太理解区别

has_resp = 有应答。即使是 servfail 这种请求不成功的应答，只要是应答，就算。
has_wanted_ans = 有应答 + 有想要的记录。比如请求 AAAA，应答必需含 AAAA 应答。

**Reply by Mrican** [reply]
感谢不吝赐教！

**Comment by wenbozzz** [comment]
@Mrican  @urlesistiana 好奇这里屏蔽qtype 65的原因是什么呢？我看到有好几个人分享的配置都屏蔽了这个type
![image](https://github.com/user-attachments/assets/34c492d4-80af-40b2-bc80-4471262c88b7)
按维基百科说的这个应该只是查找https下这个域名所对应的IP地址，通常应该跟 type A record是一样的吧？
